### PR TITLE
Additional trace information in exception messages during schema initialization

### PR DESCRIPTION
**Pull request by sungam3r**
Helps to deal with #3480

### CODE REVIEW

File: src/GraphQL/Types/Collections/SchemaTypes.cs [hunk 1]
```diff
@@ -656,7 +694,7 @@ private IGraphType AddTypeIfNotRegistered(Type type, TypeCollectionContext conte
                 else if (_builtInCustomScalars.TryGetValue(namedType, out var builtInCustomScalar))
                 {
                     foundType = builtInCustomScalar;
-                    AddType(foundType, _context);
+                    AddType(foundType, context);
```
Comment (sungam3r) [inline src/GraphQL/Types/Collections/SchemaTypes.cs:697]: @Shane32 I changed this accidentally. Looks like a bug thought. 🤔

Comment (sungam3r) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:697]: On the other side - `_context` is a field and this is the only usage outside of `Initialize` method.

Comment (Shane32) [inline src/GraphQL/Types/Collections/SchemaTypes.cs:697]: Why was this changed?  Bug?

File: src/GraphQL/GlobalSwitches.cs [hunk 2]
```diff
@@ -65,6 +65,13 @@ public static class GlobalSwitches
     /// </summary>
     public static bool UseDeclaringTypeNames { get; set; } = false;
 
+    /// <summary>
+    /// Enable this switch to see additional debugging information in exception messages during schema initialization.
+    /// <br/>
+    /// By default disabled.
+    /// </summary>
+    public static bool TrackGraphTypeInitialization { get; set; } = true;
```
Comment (Shane32) [inline src/GraphQL/GlobalSwitches.cs:73]:
```suggestion
    public static bool TrackGraphTypeInitialization { get; set; } = false;
```

Comment (Shane32) [reply src/GraphQL/GlobalSwitches.cs:73]: I don't think all this extra trace activity and metadata logging should be enabled by default.

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:697]: Accidentally. Now I puzzle over whether `_context` was used here on purpose. Awkward SchemaTypes code. I did not remember.

File: src/GraphQL/GlobalSwitches.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL/GlobalSwitches.cs:73]: I forgot to remote it 🤦‍♂️ .

File: src/GraphQL/Types/Collections/SchemaTypes.cs [hunk 3]
```diff
@@ -629,6 +665,7 @@ private void AddTypeWithLoopCheck(IGraphType resolvedType, TypeCollectionContext
             context.InFlightRegisteredTypes.Push(namedType);
             try
             {
+                using var _ = context.Trace($"AddTypeWithLoopCheck for type '{namedType.Name}'");
```
Comment (Shane32) [inline src/GraphQL/Types/Collections/SchemaTypes.cs:668]: I would like to point out that these strings get created whether tracing is enabled or not.  Either we can leave it, which is fine, or we can add the name as a second argument, like so:

```cs
IDisposable Trace(string description, string? param = null)
{
    if (tracing is enabled)
        description = param == null ? description : $"{description} '{param}'";
    // other code
}

using var _ = context.Trace("AddTypeWithLoopCheck for type", namedType.Name);
```

Comment (Shane32) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:668]: (obviously this is just an example)

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:697]: "If it ain't broke, don't fix it".... unless you're sure.

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:668]: I know, can change it.

File: src/GraphQL.Tests/Types/Collections/SchemaTypesTests.cs [hunk 4]
```diff
@@ -84,7 +84,8 @@ public void throws_exception_when_multiple_type_instances_exists()
 
         // Test
         Should.Throw<InvalidOperationException>(() => schema.Initialize())
-            .Message.ShouldBe("A different instance of the GraphType 'ObjectGraphType' with the name 'MyObject' has already been registered within the schema. Please use the same instance for all references within the schema, or use GraphQLTypeReference to reference a type instantiated elsewhere.");
+            .Message.ShouldBe(@"A different instance of the GraphType 'ObjectGraphType' with the name 'MyObject' has already been registered within the schema. Please use the same instance for all references within the schema, or use GraphQLTypeReference to reference a type instantiated elsewhere.
+To view additional trace enable GlobalSwitches.TrackGraphTypeInitialization switch.");
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/Collections/SchemaTypesTests.cs:88]: :+1:

Setting global switches in tests is failure-prone without correct setup.  If you want to do so, be sure to apply `[Collection("StaticTests")]` to the class that contains the tests, so they do not run parallel to other tests.  (This is how we fixed the intermittent culture failures.)

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:697]: Yeah, I don't know if it was intentional or not.  As such, I personally wouldn't touch it, but I might add a comment.  I'll review the remainder of this PR and leave this change for you to decide.

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 3 above)
Comment (Shane32) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:668]: If you plan to do so, then I'll review again after you do so.  Everything else looks fine.

Comment (sungam3r) [reply src/GraphQL/Types/Collections/SchemaTypes.cs]: done

File: src/GraphQL.Tests/Types/Collections/SchemaTypesTests.cs (hunk 4 above)
Comment (sungam3r) [reply src/GraphQL.Tests/Types/Collections/SchemaTypesTests.cs:88]: I would keep switch disabled.

File: src/GraphQL/Types/TypeCollectionContext.cs [hunk 5]
```diff
@@ -20,22 +20,61 @@ internal TypeCollectionContext(Func<Type, IGraphType> resolver, Action<string, I
...
+        {
+            return InitializationTrace == null
+                ? default
+                : new(this, string.Format(traceElement, arg1, arg2));
+        }
+    }
+
+    internal readonly struct TypeCollectionContextInitializationTrace : IDisposable
+    {
+        private readonly TypeCollectionContext _context;
```
Comment (Shane32) [inline src/GraphQL/Types/TypeCollectionContext.cs:67]:
```suggestion
        private readonly TypeCollectionContext? _context;
```

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/Types/Collections/SchemaTypes.cs]: I reverted.

**Review by Shane32 — APPROVED** [review]
